.circleci/config.yml

```diff
@@ -134,36 +115,6 @@ jobs:
             export KNAPSACK_PRO_CI_NODE_INDEX=$CIRCLE_NODE_INDEX
             bundle exec parallel_test -n $PARALLEL_TESTS_CONCURRENCY -e './bin/parallel_tests'
 
-      # Queue Mode: RSpec split by test examples
-      - run:
-          name: RSpec split by test cases (Queue Mode)
-          command: |
-            export KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC=$KNAPSACK_PRO_QUEUE_TEST_SUITE_TOKEN_RSPEC_SPLIT_BY_TEST_CASES
-            export KNAPSACK_PRO_RSPEC_SPLIT_BY_TEST_EXAMPLES=true
-            bundle exec rake knapsack_pro:queue:rspec
-
-      - run:
-          name: RSpec split by test cases (Regular Mode) with KNAPSACK_PRO_SLOW_TEST_FILE_PATTERN
-          command: |
-            export KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC=$KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC_SPLIT_BY_TEST_CASES
-            export KNAPSACK_PRO_RSPEC_SPLIT_BY_TEST_EXAMPLES=true
-            export KNAPSACK_PRO_SLOW_TEST_FILE_PATTERN="{spec/timecop_spec.rb}"
-            bundle exec rake knapsack_pro:rspec
-
-      # Queue Mode: Minitest
-      - run:
-          command: |
-            export KNAPSACK_PRO_RSPEC_DISABLED=true
-            export KNAPSACK_PRO_TEST_SUITE_TOKEN_MINITEST=$KNAPSACK_PRO_QUEUE_TEST_SUITE_TOKEN_MINITEST
-            bundle exec rake knapsack_pro:queue:minitest[--verbose]
-
-
-      # Regular Mode
-      - run:
-          command: |
-            export KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=false
```

Comment:
In Regular Mode [`KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=true`](https://docs.knapsackpro.com/ruby/reference/#knapsack_pro_fixed_test_suite_split-regular-mode) is a default value.

In the knapsack_pro gem we could test the following scenario:

* run build with a new CI build ID and default KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=true. It will be a new split of tests.
* test retry of CI build (similarly like we do for Queue Mode). Run build for the same CI build ID as above and use default KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=true. It should rerun the same set of tests on each CI node.

I think it's fine to remove the example here because we don't need to test `KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=false` case. It behaves the same as when you run a new CI build ID.

But if we really want to test that then we could add 3rd example after the above 2 examples:

* Run build for the same CI build ID as above builds and use KNAPSACK_PRO_FIXED_TEST_SUITE_SPLIT=false. It should do a new split of tests despite using the same CI build ID.

Reply:
It looks like it retries the same tests but in a different order? Is that correct?

https://app.circleci.com/pipelines/github/KnapsackPro/knapsack_pro-ruby/1128/workflows/f5b6681d-6e93-4f71-a221-a2e8dc1c4d28/jobs/4151/parallel-runs/1?filterBy=ALL

Reply:
It's incorrect. I expect the same set of tests to be run on retry.

I think I found the root issue. There are other CI builds running at the same time so the retry is for another build. 
We use the same branch name for matrix of Ruby versions and Minitest. Each Ruby version name should be included in the branch name so that the builds are unique and not overlapping. 

![Screenshot 2024-01-17 at 22 55 10](https://github.com/KnapsackPro/rails-app-with-knapsack_pro/assets/662545/0c293006-7f87-478b-ae79-b273125953ab)
The build distribution is the same. That's good. On retry there is a build subset ID in the response which I think was not created by the previous step in the pipeline. 
API source code:
* https://github.com/KnapsackPro/knapsack-pro-api/blob/42a80173234d8413d826395a216353b665587f59/app/controllers/api/v1/build_distributions_controller.rb#L196

To fix the issue you need to add Ruby version name to the branch:

```
export KNAPSACK_PRO_BRANCH="$CIRCLE_BRANCH--$CIRCLE_BUILD_NUM--regular--$RUBY_VERSION"
```

If the above helps, then we can do the same for RSpec to avoid similar issues. We need to add Ruby and RSpec version to the name of the branch.

Reply:
> It's incorrect. I expect the same set of tests to be run on retry.

They are the same tests, just in a different order.

> Each Ruby version name should be included in the branch name so that the builds are unique and not overlapping.

I don't think this is needed. The `$CIRCLE_BUILD_NUM` changes for each combination of the matrix.

---

I think we are good, please let me know if I'm missing anything @ArturT

Reply:
> They are the same tests, just in a different order.

Indeed. We are good.

---

My mistake I assumed the tests are executed in the same order but I was wrong. Minitest runs them in random order. Test file paths saved as 1st build subset in API are in order of execution. When we retry the build  then API returned test files in order of execution hence mismatch between 1st and 2nd API response on the screenshot. The set of tests is the same so that's good.